provider/devdocs/index.ts

```diff
@@ -0,0 +1,155 @@
+import type {
+    ItemsParams,
+    ItemsResult,
+    MentionsParams,
+    MentionsResult,
+    MetaResult,
+    Provider,
+} from '@openctx/provider'
+import Fuse from 'fuse.js'
+import { LRUCache } from 'lru-cache'
+import { parse } from 'node-html-parser'
+import { fetchDoc, fetchIndex } from './devdocs.js'
+
+/**
+ * Settings for the DevDocs OpenCtx provider.
+ */
+export type Settings = {
+    /**
+     * The list of URLs to serve.
+     *
+     * These should be top-level documentation URLs like https://devdocs.io/angular~16/ or https://devdocs.io/typescript/
+     *
+     * Additionally this supports file:// URLs for local development.
+     */
+    urls: string[]
+}
+
+/**
+ * An OpenCtx provider that fetches the content of a [DevDocs](https://devdocs.io/) entry.
+ */
+const devdocs: Provider<Settings> = {
+    meta(): MetaResult {
+        return {
+            // empty since we don't provide any annotations.
+            selector: [],
+            name: 'DevDocs',
+            features: { mentions: true },
+        }
+    },
+
+    async mentions(params: MentionsParams, settings: Settings): Promise<MentionsResult> {
+        const query = params.query?.toLowerCase()
+        if (!query) {
+            return []
+        }
+
+        const indexes = await Promise.all(settings.urls.map(url => getMentionIndex(url)))
+        const entries = indexes.flatMap(index => {
+            return index.fuse.search(query, { limit: 10 }).map(result => ({
+                score: result.score ?? 0,
+                item: result.item,
+                url: index.url,
+            }))
+        })
+
+        entries.sort((a, b) => a.score - b.score)
+
+        return entries.slice(0, 20).map(entry => {
+            return {
+                title: entry.item.name,
+                uri: new URL(entry.item.path, entry.url).toString(),
+            }
+        })
+    },
+
+    async items(params: ItemsParams): Promise<ItemsResult> {
+        const mention = params.mention
+        if (!mention) {
+            return []
+        }
+
+        const { content, hash } = await fetchDoc(mention.uri)
+        if (!hash) {
+            return [
+                {
+                    title: mention.title,
+                    url: mention.uri,
+                    ai: {
+                        content: content,
+                    },
+                },
+            ]
+        }
+
+        const root = parse(content)
+
+        let last = root.querySelector(hash)!
+        const nodes = [last]
+        while (true) {
+            const next = last.nextElementSibling
+            if (!next || next.rawTagName === 'h2') {
+                break
+            }
+            last = next
+            nodes.push(last)
+        }
+
+        return [
+            {
+                title: mention.title,
+                url: mention.uri,
+                ai: {
+                    content: nodes.map(node => node.toString()).join(''),
```

Comment:
Here I am just including raw HTML as the content. Do LLMs handle this alright, or should I do something to make it look like markdown or xml?

Reply:
In my experience, it works fine with HTML, but LLMs are generally more effective with plain text. HTML markup adds extra noise, potentially leading to less accurate responses.

Reply:
Cool, I'll stick to this. But could also steal @sqs's web provider regex / maybe use this html parser to generate simple markdown looking  output.